motoko/invoice-canister/dfx.json

```diff
@@ -1,26 +1,48 @@
 {
-    "canisters": {
-        "invoice": {
-            "dependencies": [
-                "ledger"
-            ],
-            "main": "src/invoice/main.mo",
-            "type": "motoko",
-            "declarations": {
-                "output": "test/e2e/src/declarations"
-            }
-        },
-        "ledger": {
-            "type": "custom",
-            "candid": "src/ledger/ledger.did",
-            "wasm": "src/ledger/ledger.wasm",
-            "remote": {
-                "candid": "src/ledger/ledger.did",
-                "id": {
-                    "ic": "ryjl3-tyaaa-aaaaa-aaaba-cai"
-                }
-            }
+  "canisters": {
+    "invoice": {
+      "declarations": {
+        "output": "test/e2e/src/declarations/invoice",
+        "node_compatibility": true
+      },
+      "main": "src/invoice/Invoice.mo",
+      "type": "motoko"
+    },
+    "nns-ledger": {
+      "build": "",
+      "candid": "candid/nns-ledger.did",
+      "remote": {
+        "id": {
+          "ic": "ryjl3-tyaaa-aaaaa-aaaba-cai",
+          "local": "ryjl3-tyaaa-aaaaa-aaaba-cai"
         }
+      },
+      "type": "custom",
+      "wasm": ""
+    },
+    "icp_ledger_canister": {
```

Comment:
Why is `icp_ledger` different from `nns-ledger`?

Reply:
This overlaps with the previous comment/reply. 

The answer is that when I started the work there was motivation to show how `dfx nns install` could be used to deploy a local instance of the ICP ledger canister. Integrating that led to a deployed ICP ledger whose did file conformed to the candid installed with `dfx nns import`. Coincidentally at the time I had also installed another ICP ledger with the Motoko based actor class to understand how to adapt the ICP ledger to the Invoice Canister's API. Since their dids are not entirely uniform this is what ultimately led to the `SupportedToken` variant because by that point I was figuring out how to also include the ICRC1 standard (which also had its own Motoko actor class reference added). 

After that was implemented, it seemed the best way to balance the complexity of including all those and demonstrating that the `SupportedToken` type works, was to only use `dfx nns install` along side the wasms/dids of the Rosetta repository, since less hard-coded Motoko code/files would be necessary and IIUC the Rosetta repository is (at least in part) intended for external accountability. 

Thus the two ICP ledger canisters references are different because one is for the included Rosetta wasm/did and the other for that of `dfx nns install`. Without all the background, this is explained in the beginning of [clean-startup.mjs](https://github.com/atengberg/examples/blob/ashton/invoice-bnt2/motoko/invoice-canister/clean-startup.mjs). I would have included those specifics in the readme as well, but it seemed like there was enough going on in the readme already. 

I do agree that having an example of invoice canister code that includes a "standard set" of tokens would be useful, although maybe that would only use mock classes to represent all the involved token ledger canisters. As for the main project, if it wouldn't be too much adding an SNS-1 token with `dfx sns install` seems appropriate; although a dedicated e2e/integrated example of using the invoice canister for a sns dapp also seems useful. In any case, since the scope of the project was for completing the bounty, I did not develop the `SupportedToken` type as much as I would have liked--in fact, while working on the `motoko-seller-client` example it became clear how much overlap there might be with an independent library module the invoice canister could take the form of, or for that matter the `SupportedToken` type itself. 

In any case, hopefully this code base can help move things along.

Reply:
I think we can remove the included wasm. I only was using that originally because `dfx nns install` didn't exist yet at the time

Reply:
The reason I added it back after integrating `dfx nns install` was to provide an applied demonstration of the `Local Ledger` instructions as well as have another token in showing that `SupportedToken` type works for multiple tokens (along with the E2E tests). Also while `dfx nns install` can be used to deploy an ICP ledger canister, it might not always be feasible for a project because it comes as part of all the nns canisters that also must be installed (they are omitted from `dfx.json` as they are not needed, but they are all installed in the local replica as well when `dfx nns install` is used). Additionally, it gives an opportunity to show how `zx` can be used to automate swapping the private and public dids of the ICP ledger when deploying it with the wasm. 

The motivation was to have the main project demonstrate the "main" ways an ICP (and ICRC1) ledger could be deployed while also "proving" the `SupportedToken` type can handle many entries. I had actually added both the Motoko references for the actor classes as well (for a total of 6 tokens), but that introduced some other issues, and having the wasms/dids of the Rosetta repository seemed to be the most relevant (along with `dfx nns install`) since they are intended (IUCC) for "external accountability" and designed for "plug n play" functionality. 

If it doesn't make it too complicated, adding `dfx sns install` would then complete this idea particularly now that `dfx sns install` is working. Adding it along side `dfx nns install` would definitely be additional work that I thought could be a future task along with other features such as using Candb to autoscale invoice record storage, implementing CertifiedData or encryption with threshold ECDSA, adding (a rudimentary) official "status" to invoices to do things like proper refunding designating the consolidation address other than the creator's subaccount, providing JSON output to inbound HTTP calls, or refactoring the invoice canister functionality as independent module library components that could also be used elsewhere. 

Another possibility is including a project in the examples/ subdirectory that could be tailored towards supporting a specific set of "mainnet ready" tokens (such as the mainnet ICP ledger with the ICRC1 standard as the base) that could be more readily used dropped into another project. 

Although it could be removed, I thought it'd be more helpful as an educational example to include it.